gcode/global/gozilla/gozilla.1
.\" This file is generated automatically by convert.pl from gozilla/manual.in.
.TH GOZILLA 1 "March 2010" "GNU Project"
.SH NAME
gozilla \- force mozilla to display specified part of a source file.
.SH SYNOPSIS
\fBgozilla\fP [-b browser][-p][+no] file
.br
\fBgozilla\fP [-b browser][-p] -d name
.br
.SH DESCRIPTION
\fBGozilla\fP forces \fBmozilla\fP to display specified part of a source file.
\fBGozilla\fP can be used with other browsers like firefox and epiphany.
.PP
In advance of using this command, you must execute \fBgtags\fP(1)
and \fBhtags\fP(1) at the root directory of a project to make tag files.
Then you can execute this command anywhere in the project.
.PP
First form:
.br
You can specify a source file and optional line number.
This syntax is similar to \fBvi\fP(1) and \fBemacs\fP(1).
.PP
Second form:
.br
You can specify a definition name directly. The definition name should
exist in \'GTAGS\'.
.PP
Some browsers require you to load it before executing \fBgozilla\fP.
.SH OPTIONS
The following options are available:
.TP
\fB+no\fP
Line number.
.TP
\fB-b\fP \fIbrowser\fP
Browser to use. By default, it is assumed \fBmozilla\fP.
.TP
\fB-d\fP \fIname\fP
Print object definitions.
.TP
\fB--help\fP
Show help.
.TP
\fB-p\fP
Print just a generated URL instead of displaying it.
.TP
\fIfile\fP
File name or alias name.
.TP
\fB-q\fP, \fB--quiet\fP
Quiet mode.
.TP
\fB-v\fP, \fB--verbose\fP
Verbose mode.
.TP
\fB--version\fP
Show version number.
.SH FILES
.TP
\'HTML/\'
Hypertext of source code.
.TP
\'GTAGS/\'
Tag file for object definitions.
.TP
\'$HOME/.gozillarc\'
Alias file. Please read source code for the detail.
.SH ENVIRONMENT
.TP
\fBGTAGSROOT\fP
The root directory of the project.
.TP
\fBGTAGSDBPATH\fP
The directory on which tag files exist.
This value is ignored when GTAGSROOT is not defined.
.TP
\fBBROWSER\fP
Browser to use. By default, it is assumed \fBmozilla\fP.
.SH EXAMPLES
.nf
	$ gtags
	$ htags
	$ global -x main
	main              82 ctags.c          main(argc, argv)
	$ mozilla &
	$ gozilla +82 ctags.c
	$ gozilla -d main
.PP
	$ firefox &
	$ gozilla -b firefox +82 ctags.c
.PP
.fi
.SH DIAGNOSTICS
\fBGozilla\fP exits with a non 0 value if an error occurred, 0 otherwise.
.SH "SEE ALSO"
\fBglobal\fP(1),
\fBgtags\fP(1),
\fBhtags\fP(1),
\fBfirefox\fP(1),
\fBepiphany\fP(1),
\fBmozilla\fP(1).
.PP
GNU GLOBAL source code tag system
.br
(http://www.gnu.org/software/global/).
.SH NOTES
\fBGozilla\fP means 'Global for mozilla'.
.SH BUGS
\fBGozilla\fP can accept not only source files but also text files,
directories, HTML files and even URLs, because it is omnivorous.
.SH AUTHORS
Shigio YAMAGUCHI.
.SH HISTORY
The \fBgozilla\fP command appeared in FreeBSD 2.2.2 but did not
installed by default.


NEWS
Current Version: 5.9~svn20110310

ctags-5.9~svn20110310 (17 Mar 2011)
* Added support for new "attached" and "detachable" keywords [Eiffel].
* Fixed parsing of comments after import statements and other tags, contributed by Huandari Lopez to Geany [Python].
* Fixed PHP parser to ignore keywords inside comments [PHP, Bug #1795926].
* Fixed regular expressions for Ant so they won't span multiple tags [Ant].
* Fixed infinite loop with malformed Makefiles, contributed by Martin Dorey [Make, Bug #2959889].
* Fixed Verilog parameter parsing, contributed by Nicolas Vincent [Verilog, Patch #2747828].
* Fixed error when parsing empty file [OCaml].
* Enabled Large File System support. [Bug #3062068]

ctags-5.8 (09 Jul 2009)
* Removed ".ml" as a Lisp extension (now OCaml) [Lisp].
* Added support for Ant language, contributed by David Fishburn.
* Added support for DOS Batch language, contributed by David Fishburn.
* Added support for Flex (Adobe) language, contributed by David Fishburn.
* Added support for MATLAB language, contributed by David Fishburn.
* Added support for Objective Camel (OCaml), provided by Vincent Berthoux [Patch #2738723].
* Added support for TeX language, contributed by David Fishburn.
* Added support for VHDL language, contributed by Nicolas Vincent [Bug #1943306].
* Added support for Pyrex/Cython declarations [Python].
* Added support for "v" kind, for variables [Python].
* Added support for class and member variables [PHP, Bug #1037086].
* Added support for recent enhancements to Eiffel language [Eiffel].
* Added support for ASP classes, contributed by Zendhi Nagao; changes meaning of 'c' kind flag [ASP].
* Added regex support when compiling with MinGW. Gnu regex module now included in all distributions.
* Fixed detection of triple strings inside other strings [Python, Bug #1988130].
* Fixed an endless loop with comments in triple strings [Python, Bug #1988027].
* Fixed bug where functions were sometimes seen as methods [Python, Bug #1988026].
* Fixed parsing of method parameter annotations, fix contributed by Paolo "blaisorblade" Giarrusso [Java, Bug #2049723, #2117073].
* Fixed parsing of global scope qualifiers in base class lists [C++, Bug #1799343].
* Fixed bug where namespace members were given kinds corresponding to globals [C++, Bug #1924919, #1575055].
* Fixed parsing of "else" [C#, Bug #1830344].
* Fixed parsing of derived enums [C#, Bug #1515910].
* Fixed parsing of "foreach" [C#, Bug #1830343].
* Fixed parsing of simple generic classes [C#, Bug #1515910].
* Fixed bug with detecting identifiers inside variables [Python, Bug #1809024].
* Fixed bug with detecting identifiers at the start of variables [Python, Bug #1856363].
* Fixed parsing of triple single-quoted multi-line strings [Python, Bug #1906062].
* Changed to newer version of autoconf, changing configure.in to configure.ac.

ctags-5.7 (04 Sep 2007)
* Added support for DIM AS [Freebasic, Bug #1741778].
* Added support for arbitrary nesting depth [Python, Bug #1684786, Debian bug #409078].
* Added support for verbatim string literals [C#, Bug #1515910].
* Added support for .ctags as well as ctags.cnf on Windows [Bug #1246506].
* Added support for non-extern, non-static functions returning wchar_t, contributed by Aaron Peromsik [C++, Patch #1458930].
* Added support for numerous revision control systems including Bazaar and Mercurial [Bug #1472894].
* Added support for enums [Java, Bug #1730485, Bug #1517143, Patch #1027395, Patch #1528507].
* Added support for multiple-level namespace declarations [C#].
* Added .svn to list of directories ignored during recursion (--recurse).
* Added support for BlitzBasic, PureBasic and FreeBasic [FR #1100506].
* Added support for interfaces and static/public/protected/private functions [PHP].
* Added support for 'package' keyword [Perl].
* Added support for multi-line subroutine, package, and constant definitions [Perl].
* Added support for optional subroutine declarations [Perl].
* Added support for formats [Perl].
* Added support for new convert keyword [Eiffel].
* Added optional tags for forward variable declarations (e.g. 'struct C;') [C, C++].
* Changed parsing of option input file (-L) to strip trailing white space.
* Ignore comments mixed into definitions and declarations [Perl].
* Fixed detecting labels with whitespace after label name [Perl, Bug #1752361]
* Fixed parsing of generic classes/interfaces [Java, Bug #1447756].
* Fixed misidentification of fully qualified function calls as labels [Perl].
* Fixed parsing of inner classes [Python, Bug #1411963].
* Fixed line continuation [Python, Bug #928001, Patch #819471].
* Fixed parsing of annotations [Java, Bug #1691412].
* Fixed block-comment parsing [Verilog, Patch #1458042, Bugs #960316, #1111214, #1606569, #1615060].
* Fixed typo in man page [Debian bug #366412].
* Fixed missing chunk of text in man page and over-use of hyphens in UTF-8 locales [Debian bug #271323].
* Fixed parsing of ` as a method name [Ruby].
* Fixed parsing of keywords in string literals [Ruby, Bug #1742588].
* Fixed potential segmentation violation [Bug #1672834, Bug #1222926].
* Fixed parsing of destructors with whitespace after the '~' [C++, Bug #1585745].
* Fixed default access of unions to be public [C++, Bug #1548443].
* Fixed various memory leaks, mostly contributed by Dmitry Antipov.
* Fixed parsing of `define [Verilog, Bug #961001].
* Fixed crashes involving '/' [Verilog, Bug #1743330].
* Fixed compilation problem on MinGW [Bug #1517424].
* Fixed generation of HTML-formatted man page [Bug #1645864].
* Fixed recognition of Python scripts having '#!/usr/bin/python' as first line [Bug #1764148].
* Fixed parsing of Fortran comment-to-end-of-line with no newline before EOF [Debian bug #432872].
* Fixed parsing of << [C/C++, Bugs #1020715, #1093123, #1770479, #1770607].
* Fixed parsing of fully-qualified type names [Java, Bug #814263].
* Fixed handling of lone carriage-return characters in file [Bug #1773926].

ctags-5.6 (Mon May 29 2006)

gcode/global/gozilla/gozilla.c
/*
 * Copyright (c) 1997, 1998, 1999, 2000, 2002, 2003, 2004, 2006
 *	Tama Communications Corporation
 *
 * This file is part of GNU GLOBAL.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <shellapi.h>
/*
 * Don't remove the following code which seems meaningless.
 * Since WIN32 has another SLIST_ENTRY, we removed the definition
 * so as not to cause the conflict.
 */
#ifdef SLIST_ENTRY
#undef SLIST_ENTRY
#endif
#endif

#include "global.h"
#include "regex.h"
#include "const.h"

static void usage(void);
static void help(void);

const char *gozillarc = ".gozillarc";
#ifdef __DJGPP__
const char *dos_gozillarc = "_gozillarc";
#endif
STRHASH *sh;

static void load_alias(void);
static const char *alias(const char *);
int main(int, char **);
void getdefinitionURL(const char *, STRBUF *);
void getURL(const char *, STRBUF *);
int isprotocol(const char *);
int convertpath(const char *, const char *, const char *, STRBUF *);
void makefileurl(const char *, int, STRBUF *);
void show_page_by_url(const char *, const char *);
#ifndef isblank
#define isblank(c)	((c) == ' ' || (c) == '\t')
#endif

char cwd[MAXPATHLEN];
char root[MAXPATHLEN];
char dbpath[MAXPATHLEN];

int bflag;
int pflag;
int qflag;
int Cflag;
int vflag;
int show_version;
int linenumber = 0;
int debug;

static void
usage(void)
{
	if (!qflag)
		fputs(usage_const, stderr);
	exit(2);
}
static void
help(void)
{
	fputs(usage_const, stdout);
	fputs(help_const, stdout);
	exit(0);
}

/*
 * load_alias: load alias value.
 *
 * [$HOME/.gozillarc]
 * +-----------------------
 * |a:http://www.gnu.org
 * |f = file:/usr/share/xxx.html
 * |www	http://www.xxx.yyy/
 */
static void
load_alias(void)
{
	FILE *ip;
	STRBUF *sb = strbuf_open(0);
	char *p;
	int flag = STRBUF_NOCRLF;
	struct sh_entry *ent;

	sh = strhash_open(10);
	if (!(p = get_home_directory()))
		goto end;
	if (!test("r", makepath(p, gozillarc, NULL)))
#ifdef __DJGPP__
		if (!test("r", makepath(p, dos_gozillarc, NULL)))
#endif
			goto end;
	if (!(ip = fopen(makepath(p, gozillarc, NULL), "r")))
#ifdef __DJGPP__
		if (!(ip = fopen(makepath(p, dos_gozillarc, NULL), "r")))
#endif
			goto end;
	while ((p = strbuf_fgets(sb, ip, flag)) != NULL) {
		char *name, *value;

		flag &= ~STRBUF_APPEND;
		if (*p == '#')
			continue;
		if (strbuf_unputc(sb, '\\')) {
			flag |= STRBUF_APPEND;
			continue;
		}
		while (*p && isblank(*p))	/* skip spaces */
			p++;
		name = p;
		while (*p && isalnum(*p))	/* get name */
			p++;
		*p++ = 0;
		while (*p && isblank(*p))	/* skip spaces */
			p++;
		if (*p == '=' || *p == ':') {
			p++;
			while (*p && isblank(*p))/* skip spaces */
				p++;
		}
		value = p;
		while (*p && !isblank(*p))	/* get value */
			p++;
		*p = 0;
		ent = strhash_assign(sh, name, 1);
		if (ent->value)
			(void)free(ent->value);
		ent->value = check_strdup(value);
	}
	fclose(ip);
end:
	strbuf_close(sb);
}
/*
 * alias: get alias value.
 *
 *	i)	alias_name	alias name
 *	r)			its value
 */
static const char *
alias(const char *alias_name)
{
	struct sh_entry *ent = strhash_assign(sh, alias_name, 0);
	return ent ? ent->value : NULL;
}

/*
 * locate_HTMLdir: locate HTML directory made by htags(1).
 *
 *	r)		HTML directory
 */
static const char *
locate_HTMLdir(void)
{
	static char htmldir[MAXPATHLEN];

	if (test("d", makepath(dbpath, "HTML", NULL)))
		strlimcpy(htmldir, makepath(dbpath, "HTML", NULL), sizeof(htmldir));
	else if (test("d", makepath(root, "HTML", NULL)))
		strlimcpy(htmldir, makepath(root, "HTML", NULL), sizeof(htmldir));
	else if (test("d", makepath(root, "html/HTML", NULL)))
		/* Doxygen makes HTML in doxygen's html directory. */
		strlimcpy(htmldir, makepath(root, "html/HTML", NULL), sizeof(htmldir));
	else
		die("hypertext not found. See htags(1).");
	if (vflag)
		fprintf(stdout, "HTML directory '%s'.\n", htmldir);
	return (const char *)htmldir;
}
int
main(int argc, char **argv)
{
	char c;
	const char *p, *browser = NULL, *definition = NULL;
	STRBUF *arg = strbuf_open(0);
	STRBUF *URL = strbuf_open(0);

	while (--argc > 0 && ((c = (++argv)[0][0]) == '-' || c == '+')) {
		if (argv[0][1] == '-') {
			if (!strcmp("--help", argv[0]))
				help();
			else if (!strcmp("--version", argv[0]))
				show_version++;
			else if (!strcmp("--quiet", argv[0])) {
				qflag++;
				vflag = 0;
			} else if (!strcmp("--verbose", argv[0])) {
				vflag++;
				qflag = 0;
			} else
				usage();
			continue;
		}
		if (c == '+') {
			linenumber = atoi(argv[0] + 1);
			continue;
		}
		p = argv[0] + 1;
		switch (*p) {
		case 'b':
			browser = argv[1];
			--argc; ++argv;
			break;
		case 'd':
			definition = argv[1];
			--argc; ++argv;
			break;
		case 'p':
			pflag++;
			break;
		case 'q':
			qflag++;
			setquiet();
			break;
		case 'v':
			vflag++;
			break;
		default:
			usage();
		}
	}
	if (show_version)
		version(progname, vflag);
	/*
	 * Load aliases from .gozillarc.
	 */
	load_alias();

	/*
	 * Decide browser.
	 */
	if (!browser && getenv("BROWSER"))
		browser = getenv("BROWSER");
	if (!browser && alias("BROWSER"))
		browser = alias("BROWSER");
	if (!browser)
		browser = "mozilla";

	/*
	 * Replace alias name.
	 */
	if (definition == NULL) {
		if (argc == 0)
			usage();
		strbuf_puts(arg, argv[0]);
		/*
		 * Replace with alias value.
		 */
		if ((p = alias(strbuf_value(arg))) != NULL) {
			strbuf_reset(arg);
			strbuf_puts(arg, p);
		}
	}
	/*
	 * Get URL.
	 */
	if (!definition && isprotocol(strbuf_value(arg))) {
		strbuf_puts(URL, strbuf_value(arg));
	} else {
		getdbpath(cwd, root, dbpath, 0);
		if (definition)
			getdefinitionURL(definition, URL);
		else
			getURL(strbuf_value(arg), URL);
	}
	if (pflag) {
		fprintf(stdout, "%s\n", strbuf_value(URL));
		if (vflag)
			fprintf(stdout, "using browser '%s'.\n", browser);
		exit(0);
	}
	/*
	 * Show URL's page.
	 */
	show_page_by_url(browser, strbuf_value(URL));
	exit(0);
}

/*
 * getdefinitionURL: get URL includes specified definition.
 *
 *	i)	arg	definition name
 *	o)	URL	URL begin with 'file:'
 */
void
getdefinitionURL(const char *arg, STRBUF *URL)
{
	FILE *fp;
	char *p;
	SPLIT ptable;
	int status = -1;
	STRBUF *sb = strbuf_open(0);
	const char *htmldir = locate_HTMLdir();
	const char *path = makepath(htmldir, "MAP", NULL);

	if (!test("f", path))
		die("'%s' not found. Please reconstruct hypertext using the latest htags(1).", path);
	fp = fopen(path, "r");
	if (!fp)
		die("cannot open '%s'.", path);
	while ((p = strbuf_fgets(sb, fp, STRBUF_NOCRLF)) != NULL) {
		if (split(p, 2, &ptable) != 2)
			die("illegal format.");
		if (!strcmp(arg, ptable.part[0].start)) {
			status = 0;
			break;
		}
	}
	fclose(fp);
	if (status == -1)
		die("definition %s not found.", arg);
	strbuf_reset(URL);
	/*
	 * convert path into URL.
	 */
	makefileurl(makepath(htmldir, ptable.part[1].start, NULL), 0, URL);
	recover(&ptable);
	strbuf_close(sb);
}
/*
 * getURL: get URL of the specified file.
 *
 *	i)	file	file name
 *	o)	URL	URL begin with 'file:'
 */
void
getURL(const char *file, STRBUF *URL)
{
	char *p;
	char buf[MAXPATHLEN];
	STRBUF *sb = strbuf_open(0);
	const char *htmldir = locate_HTMLdir();

	if (!test("f", file) && !test("d", file))
		die("path '%s' not found.", file);
	p = normalize(file, get_root_with_slash(), cwd, buf, sizeof(buf));
	if (convertpath(dbpath, htmldir, p, sb) == 0)
		makefileurl(strbuf_value(sb), linenumber, URL);
	else
		makefileurl(realpath(file, buf), 0, URL);
	strbuf_close(sb);
}
/*
 * isprotocol: return 1 if url has a procotol.
 *
 *	i)	url	URL
 *	r)		1: protocol, 0: file
 */
int
isprotocol(const char *url)
{
	const char *p;

	if (locatestring(url, "file:", MATCH_AT_FIRST))
		return 1;
	/*
	 * protocol's style is like http://xxx.
	 */
	for (p = url; *p && *p != ':'; p++)
		if (!isalnum(*p))
			return 0;
	if (!*p)
		return 0;
	if (*p++ == ':' && *p++ == '/' && *p == '/')
		return 1;
	return 0;
}
/*
 * convertpath: convert source file into hypertext path.
 *
 *	i)	dbpath	dbpath
 *	i)	htmldir	HTML directory made by htags(1)
 *	i)	path	source file path
 *	o)	sb	string buffer
 *	r)		0: normal, -1: error
 */
int
convertpath(const char *dbpath, const char *htmldir, const char *path, STRBUF *sb)
{
	static const char *suffix[] = {".html", ".htm"};
	static const char *gz = ".gz";
	int i, lim = sizeof(suffix)/sizeof(char *);
	const char *p;

	strbuf_reset(sb);
	strbuf_puts(sb, htmldir);
	strbuf_puts(sb, "/S/");
	/*
	 * new style.
	 */
	if (gpath_open(dbpath, 0) == 0) {
		int tag1 = strbuf_getlen(sb);

		p = gpath_path2fid(path, NULL);
		if (p == NULL) {
			gpath_close();
			return -1;
		}
		gpath_close();
		strbuf_puts(sb, p);
		for (i = 0; i < lim; i++) {
			int tag2 = strbuf_getlen(sb);
			strbuf_puts(sb, suffix[i]);
			if (test("f", strbuf_value(sb)))
				return 0;
			strbuf_puts(sb, gz);
			if (test("f", strbuf_value(sb)))
				return 0;
			strbuf_setlen(sb, tag2);
		}
		strbuf_setlen(sb, tag1);
	}
	/*
	 * old style.
	 */
	for (p = path + 1; *p; p++)
		strbuf_putc(sb, (*p == '/') ? ' ' : *p);
	for (i = 0; i < lim; i++) {
		int tag = strbuf_getlen(sb);
		strbuf_puts(sb, suffix[i]);
		if (test("f", strbuf_value(sb)))
			return 0;
		strbuf_puts(sb, gz);
		if (test("f", strbuf_value(sb)))
			return 0;
		strbuf_setlen(sb, tag);
	}
	return -1;
}
/*
 * makefileurl: make url which start with 'file:'.
 *
 *	i)	path	path name (absolute)
 *	i)	line	!=0: line number
 *	o)	url	URL
 *
 * makefileurl('/dir/a.html', 10)   => 'file:///dir/a.html#L10'
 *
 * (Windows32 environment)
 * makefileurl('c:/dir/a.html', 10) => 'file://c|/dir/a.html#L10'
 */
void
makefileurl(const char *path, int line, STRBUF *url)
{
	strbuf_puts(url, "file://");
#if _WIN32 || __DJGPP__
	/*
	 * copy drive name. (c: -> c|)
	 */
	if (isalpha(*path) && *(path+1) == ':') {
		strbuf_putc(url, *path);
		strbuf_putc(url, '|');
		path += 2;
	}
#endif
	strbuf_puts(url, path);
	if (line) {
		strbuf_puts(url, "#L");
		strbuf_putn(url, line);
	}
}
/*
 * show_page_by_url: show page by url
 *
 *	i)	browser browser name
 *	i)	url	URL
 */
#if defined(_WIN32)
/* Windows32 version */
void
show_page_by_url(const char *browser, const char *url)
{
	if (ShellExecute(NULL, NULL, browser, url, NULL, SW_SHOWNORMAL) <= (HINSTANCE)32)
		die("Cannot load %s (error = 0x%04x).", browser, GetLastError());
}
#elif defined(__DJGPP__)
/* DJGPP version */
void
show_page_by_url(const char *browser, const char *url)
{
	char com[MAXFILLEN];
	char *path;

	/*
	 * assume a Windows browser if it's not on the path.
	 */
	if (!(path = usable(browser))) {
		/*
		 * START is an internal command in XP, external in 9X.
		 */
		if (!(path = usable("start")))
			path = "cmd /c start";
		snprintf(com, sizeof(com), "%s %s \"%s\"", path, browser, url);
	} else {
		snprintf(com, sizeof(com), "%s \"%s\"", path, url);
	}
	system(com);
}
#else
/* UNIX version */
void
show_page_by_url(const char *browser, const char *url)
{
	char com[1024];

	/*
	 * Browsers which have openURL() command.
	 */
	if (locatestring(browser, "mozilla", MATCH_AT_LAST) ||
	    locatestring(browser, "netscape", MATCH_AT_LAST) ||
	    locatestring(browser, "netscape-remote", MATCH_AT_LAST))
	{
		char *path;

		if (!(path = usable(browser)))
			die("%s not found in your path.", browser);
		snprintf(com, sizeof(com), "%s -remote \"openURL(%s)\"", path, url);
	}
	/*
	 * Generic browser.
	 */
	else {
		snprintf(com, sizeof(com), "%s \"%s\"", browser, url);
	}
	system(com);
}
#endif
